leaf-server/src/main/java/org/dreeam/leaf/util/list/OptimizedNonNullListArrayList.java

```diff
@@ -0,0 +1,428 @@
+    private transient int cachedHashCode = 0;
+    private transient boolean hashValid = false;
+
+    // === CONSTRUCTORS ===
+
+    public OptimizedNonNullListArrayList(int capacity, @Nullable E defaultValue) {
+        if (capacity < 0) throw new IllegalArgumentException("Initial capacity (" + capacity + ") is negative");
+
+        // round up to next power of 2 for better cache perf
+        capacity = nextPowerOfTwo(capacity);
+
+        if (capacity == 0) {
+            this.a = (E[])ObjectArrays.EMPTY_ARRAY;
+        } else {
+            this.a = (E[])new Object[capacity];
+        }
+
+        this.defaultValue = defaultValue;
+        this.flags = defaultValue != null ? HAS_DEFAULT_FLAG : 0;
+    }
+
+    public OptimizedNonNullListArrayList(@Nullable E defaultValue) {
+        this.a = (E[])ObjectArrays.DEFAULT_EMPTY_ARRAY;
+        this.defaultValue = defaultValue;
+        this.flags = defaultValue != null ? HAS_DEFAULT_FLAG : 0;
+    }
+
+    public OptimizedNonNullListArrayList(final E[] sourceArray, @Nullable E defaultValue) {
+        this(sourceArray.length, defaultValue);
+        System.arraycopy(sourceArray, 0, this.a, 0, sourceArray.length);
+        this.size = sourceArray.length;
+    }
+
+    // === BIT MANIPULATION UTILITIES ===
+
+    /**
+     * Fast power-of-2 calculation
+     */
+    private static int nextPowerOfTwo(int n) {
+        if (n <= 1) return 1;
+        if ((n & (n - 1)) == 0) return n; // Already power of 2
+
+        n--;
+        n |= n >>> 1;
+        n |= n >>> 2;
+        n |= n >>> 4;
+        n |= n >>> 8;
+        n |= n >>> 16;
+        return ++n;
+    }
+
+    /**
+     * Check if number is power of 2
+     */
+    private static boolean isPowerOfTwo(int n) {
+        return n > 0 && (n & (n - 1)) == 0;
+    }
+
+    // === FLAG UTILITIES ===
+
+    private boolean hasFlag(int flag) {
+        return (flags & flag) != 0;
+    }
+
+    // === GROWTH LOGIC ===
+
+    private void grow(int minCapacity) {
+        final int oldCapacity = a.length;
+        if (minCapacity <= oldCapacity) return;
+
+        int newCapacity;
+
+        if (a == ObjectArrays.DEFAULT_EMPTY_ARRAY) {
+            newCapacity = Math.max(DEFAULT_INITIAL_CAPACITY, minCapacity);
+        } else {
+            // use shifts instead of division/multiplication
+            if (oldCapacity < CAPACITY_THRESHOLD) {
+                // For smaller arrays, use 1.5x growth
+                newCapacity = oldCapacity + (oldCapacity >>> GROWTH_SHIFT);
+            } else {
+                // For larger arrays, use more conservative growth to avoid OOM
+                newCapacity = oldCapacity + (oldCapacity >>> 2); // 1.25x growth
+            }
+
+            newCapacity = Math.max(newCapacity, minCapacity);
+
+            if (newCapacity > MAX_ARRAY_SIZE) {
+                newCapacity = hugeCapacity(minCapacity);
+            }
+            if (newCapacity < 1024 && !isPowerOfTwo(newCapacity)) {
```

Comment:
```suggestion
            if (newCapacity < 1024) {
```

Seems not needed to check `isPowerOfTwo` when `nextPowerOfTwo` will just return the input in that case.